scrapy_google_sheets_exporter/gsheets_exporter.py

```diff
@@ -0,0 +1,82 @@
+import csv
+import logging
+from io import TextIOWrapper
+
+import gspread
+from scrapy.exceptions import NotConfigured
+from scrapy.extensions.feedexport import BlockingFeedStorage, build_storage
+
+logger = logging.getLogger(__name__)
+
+
+class GoogleSheetsFeedStorage(BlockingFeedStorage):
+    def __init__(self, uri, credentials, *, feed_options=None):
+        if credentials:
+            self.gc = gspread.service_account_from_dict(credentials)
+        else:
+            raise NotConfigured(
+                "Must specify GOOGLE_CREDENTIALS (dict) in the spider settings."
+            )
+
+        self.spreadsheet_key, self.sheet_name = self.parse_gsheets_uri(uri)
+        self.spreadsheet = self.gc.open_by_key(self.spreadsheet_key)
+        self.sheet = self.spreadsheet.worksheet(self.sheet_name)
+        self.feed_options = feed_options or {}
+        self.overwrite = self.feed_options.get("overwrite", False)
+        self.fields = self.feed_options.get("fields", [])
+        self.format = self.feed_options.get("format", "csv")
+
+        if self.format != "csv":
+            raise NotConfigured(
+                "This feed exporter only supports csv format. "
+                f"Please update the FEEDS settings by replacing {self.format} format."
+            )
+
+    @classmethod
+    def from_crawler(cls, crawler, uri, *, feed_options=None):
+        return build_storage(
+            cls,
+            uri,
+            credentials=crawler.settings["GOOGLE_CREDENTIALS"],
+            feed_options=feed_options,
+        )
+
+    def _store_in_thread(self, file):
+        file.seek(0)
+        csv_data = csv.reader(
+            TextIOWrapper(file, newline="\r\n")
+        )
+        data_header = next(csv_data)
+
+        header = self.fields or data_header
+        if self.overwrite:
+            logger.warning("FEED option 'overwrite' was set to True.")
+            self.sheet.clear()
+            self.sheet.append_row(header)
+
+        else:
+            if self.sheet.row_values(1):
+                header = self.sheet.row_values(1)
+                logger.warning(
+                    "FEED option 'overwrite' was set to False. Since we are appending to"
+                    f"existing data, only the following fields will be exported: {header}."
+                )
+            else:
+                self.sheet.append_row(header)
+
+        rows = []
+        for line in csv_data:
+            row = dict(zip(data_header, line))
+            rows.append([v for k, v in row.items() if k in header])
```

Comment:
What if the fields are in a different order in the spreadsheet header and in Scrapy? Will that not result in a mess?

Upstream Scrapy takes a simpler, more naive approach:
- You do not omit the header if overwriting a document with rows, instead you give an option top omit the header, and make it `False` by default. Users that want to omit the header must set it to true.
- You use the fields as they come from the written CSV file, you do not attempt to adjust them to the existing header. If they do not match the existing headers, it is up to the user to make sure `fields_to_export` matches the headers in the spreadsheet.

Scrapy does so probably because reading the first row of the existing document to determine the existing header is not always possible (e.g. write-only storages).

I am OK with either approach:
- We can go smart, but then we need to make it reliable, e.g. take into account header order.
- We can go naive, provide a setting to skip the header and simply append rows regardless of existing content.

Regardless of the approach, I wonder if we could extract some of this functionality into a function, to be able to add some automated tests for it, i.e. something that would have caught the order mismatch issue.

Reply:
@Gallaecio I agree that we should follow the naive approach. Does this setting to skip the header already exist?  I can see that https://docs.scrapy.org/en/2.3/topics/exporters.html#csvitemexporter contains an `include_headers_line` parameter, but is this an option the user can set in the spider settings?

Reply:
> is this an option the user can set in the spider settings?

There is an `item_export_kwargs` [feed option](https://docs.scrapy.org/en/latest/topics/feed-exports.html#feed-options) that can be used for this.

Reply:
@Gallaecio I made the changes discussed, please take a look. I explained the use of include_headers_line and other feed options in the README please let me know if you think it's clear enough.